### ISSUE TITLE
cannot get right result when training  

### ISSUE THREAD

**Issue by hexiangquan** [eriklindernoren/PyTorch-YOLOv3#2]

**Comment by eriklindernoren** [comment]
Hi,

How long did you train the model? There is still work to be done on the training support. Data augmentation as well as weight and learning rate decays needs to be added as well as some other stuff.

**Comment by PiggyGenius** [comment]
Can you please give more details about what still needs to be done for training to work ? I am currently working on your training code and would like to implement what is missing.

**Comment by eriklindernoren** [comment]
The confidence mask needs to be fixed. I will probably get around to fixing that pretty soon. The issue I keep running into is that the recall and precision seems to decrease as I train the model but during inference (using `model.eval()`) the model outputs junk. Maybe this can be attributed to the way the network is trained w.r.t. the confidence loss at the moment but I'm not sure. I'll keep experimenting. Feel free to do the same if you want. :)

**Comment by glenn-jocher** [comment]
I see a similar dissociation between good training results but poor inference later on. 

I don't see an obvious culprit yet. One small change I noticed is that the loss criterion for xy coordinates should be MSE rather than BCE.

I think the selection of the best anchor in `build_targets` has a bug in it also, as it relies on iou's based on a common top-left corner. I fixed this and vectorized these operations in my cloned repo, and see slightly different results. I could submit a pull request if you'd like to take a look.

The code is very clean and concise though, nice work.

**Comment by libzzluo** [comment]
@glenn-jocher Hi.. Would you mind sharing your code? I met the same problem, but i can't fix it. Thanks

**Comment by glenn-jocher** [comment]
@libzzluo of course. My update to @eriklindernoren 's yolo v3 is available here. I'm not sure how usable it is to you out of the box however, as I've significantly modified it to not just fix the convergence issue but also to adapt it to the xView 2018 Object Detection Challenge (rather than COCO).

This repo converges on training with the xView dataset, and was able to produce a 0.16 mAP in the challenge (highest mAP was .27). I attached a picture of the losses, precision and recall during training here. I'm going to be branching this repo into a new COCO-specific repo in the coming days, I'll post that when complete.

https://github.com/ultralytics/xview-yolov3

![xview_training_loss](https://user-images.githubusercontent.com/26833433/44572023-db33eb00-a782-11e8-981d-a65625fad5b2.png)

**Comment by feidongxi** [comment]
THANKS!!! @glenn-jocher That's amazing!!! I will try the code on my repo.😊😊

**Comment by glenn-jocher** [comment]
Ah @feidongxi @libzzluo yes you can try and fork this right now, but it might be easier if you wait a day or two, as I'm going to backtrack all the COCO -> xView specific changes I made into a new repo so that it's back to the COCO facing implementation Erik originally created. In any case, I'll post that when its done. 

@eriklindernoren I could submit a pull request at that point if you'd like to try to merge these changes, but first of course I need to verify that COCO converges to a proper mAP, which will take more time, perhaps a week.

**Comment by eriklindernoren** [comment]
@glenn-jocher That sounds great. Haven't had much time to work on this lately. Appreciate it!

**Comment by glenn-jocher** [comment]
@eriklindernoren got it. What happened when you tried to train originally?

I've validated your .58 mAP in my forked repo after realigning it to COCO (using Redmon's weights), and I've tentatively trained a few epochs from scratch with good convergence, but I'm realizing it will be a significant challenge replicating his mAP after 160 epochs simply due to a few missing details in his paper, such as his polynomial learning rate scheduler, his multi-scale training, and his augmentation strategies, which are touched on but not explicitly described. 

Some or all of these may be in the darknet repo, do you have any info on them? If not I can try and take my best stab at it and see where the mAP lands. Theoretically my update is capable of all these things, including full augmentation, so its just a matter of figuring out what settings to use.

It's also going to take longer than I thought. My 1080 Ti appears to be good for about 16 epochs per day, so I imagine ~10 days to get to 160 epochs.

**Comment by libzzluo** [comment]
@glenn-jocher 
I check the output log file built by origin darknet framework, the network randomly change its input size between 320 and 608 (the step is 32) every 10 epochs. What's more, if i set the `random = 1` in the origin framework's `yolov3.cfg`, the GPU (1080) memory (8G) will be insufficient sometimes. 

I will also try to analyze the details in the source code and read the paper. If anything is discovered, I will add it in time. But the C source code is sooo complicated...

**Comment by glenn-jocher** [comment]
I created examples with and without augmentation to illustrate the results. Both spatial and colorspace augmentation are active here. I toned down the rotation to +/- 5 deg and added random shear of +/- 3 deg (both of these can be disabled). 

Augmented:
![augmented](https://user-images.githubusercontent.com/26833433/44629564-885c5e00-a951-11e8-8172-f409791f861a.jpg)

Standard
![default](https://user-images.githubusercontent.com/26833433/44629565-885c5e00-a951-11e8-95e2-bbd6a7f5028a.jpg)

**Comment by glenn-jocher** [comment]
I've finished creating my updated yolov3 repository:
https://github.com/ultralytics/yolov3

I've started training COCO using this repo, including the augmentation shown above. I'm running about 15 epochs per day so it will take a while to get to 160. Current progress at 25 epochs is pictured below. 

One concern I have, besides my stated unknowns above is that I see significant improvement in precision and recall using CELoss in place of BCELoss for the classification term. This is contrary to the original yolov3 loss function, which uses BCE for both classification and objectness, and MSE on the bounding boxes. This could be an indicator of underlying problems elsewhere. Unfortunately I suppose I have to wait another week until I reach epoch 160 to see what the final effect of this change is.

![figure_1](https://user-images.githubusercontent.com/26833433/44945324-38651780-ade7-11e8-9cd5-44969a55e510.png)

**Comment by lianuo** [comment]
@glenn-jocher thanks for sharing the training code

**Comment by eriklindernoren** [comment]
Awesome @glenn-jocher. Could you make a PR with your additions?

**Comment by eriklindernoren** [comment]
Yes, I believe there is an issue with `Darknet.save_weights` (see #89). I have changed to saving and loading state dicts in master and I now I get the model to converge, as well as preserved performance when saving and loading the model.
